### PR TITLE
Added datapackages support

### LINKED ISSUE #17: Add support for datapackages
# Overview

Current task configuration/descriptor doesn't support datapackages because of structure:
```
files:
settings:
```

# Ideas

We could introduce attribute `datapackages` (singular or plural?) wich will be mutually exclusive(?) with `files`:
```
datapackages:
settings:
```

If there will be more than 1 datapackages it will require a few tasks inside one build (based on current `goodtables` capabilities).

# Tasks

- [ ] add `datapackages` key to `goodtables.yml`
- [x] implement support on goodtables.py level - https://github.com/frictionlessdata/goodtables-py/issues/153
- [ ] tests

### PR DESCRIPTION
**Pull request by roll**
- fixes #17
- improved `goodtables.utils.jobconf` API 
  - encapsulated YAML parsing
  - sync with `goodtables-py`
  - and more minor
- recovered POST /api/job

### CODE REVIEW

File: goodtablesio/tasks/validate.py [hunk 1]
```diff
@@ -29,28 +29,17 @@ def validate(validation_conf, job_id):
...
-    # Safety checks
-    if not validation_conf.get('files'):
-        params = {
-            'id': job_id,
-            'finished': datetime.datetime.utcnow(),
-            'status': 'error',
-            'error': {'message': 'No files to validate'}
-        }
-
-    else:
```
Comment (roll) [inline goodtablesio/tasks/validate.py]: This now covered in `validation-conf.yml`

File: goodtablesio/tasks/validate.py [hunk 2]
```diff
@@ -29,28 +29,17 @@ def validate(validation_conf, job_id):
         }
         models.job.update(params)
 
-    # Safety checks
-    if not validation_conf.get('files'):
```
Comment (roll) [inline goodtablesio/tasks/validate.py]: This now covered in `validation-conf.yml`

**Comment by roll** [comment]
@amercader 
It's ready for review. A little bit noisy but I suppose I've done a good job cleaning our `utils.jobconf` API and concepts (including sync with `goodtables-py` updated API).

File: goodtablesio/integrations/s3/tasks/jobconf.py [hunk 3]
```diff
@@ -36,32 +36,29 @@ def get_validation_conf(bucket, job_id):
...
-            Bucket=bucket_name, Key='goodtables.yml')
-
-        job_conf = parse_job_conf(yml_file['Body'].read())
-
-    except botocore.exceptions.ClientError as e:
-        if e.response['Error']['Code'] == 'NoSuchKey':
-            job_conf = {'files': '*'}
+        yml_file = client.get_object(Bucket=bucket_name, Key='goodtables.yml')
+        job_conf_text = yml_file['Body'].read()
+    except botocore.exceptions.ClientError as exception:
```
Comment (amercader) [inline goodtablesio/integrations/s3/tasks/jobconf.py]: Add the `if e.response['Error']['Code'] == 'NoSuchKey'` bit so we don't catch other exceptions

File: goodtablesio/utils/jobconf.py [hunk 4]
```diff
@@ -119,25 +59,75 @@ def make_validation_conf(job_files, job_conf, job_base=None):
                         and job_base):
                     item['schema'] = '/'.join(
                         [job_base, item['schema']])
-                validation_conf['files'].append(item)
-
-    # Copy settings
+                validation_conf['source'].append(item)
+
+    # Datapackages
+    if 'datapackages' in job_conf:
```
Comment (amercader) [inline goodtablesio/utils/jobconf.py]: So in theory we could have configurations with both `files`  and `datapackages`, wouldn't these cause issues? eg files could get validated twice? Perhaps is best if both are exclusive, ie one or the other?

**Comment by pwalsh** [comment]
Maybe we can have order be important: if a file is in a datapackage, don't also handle it in files. Simple and easy to document as an FAQ or something.

**Comment by roll** [comment]
About `datapackage/files`:
- yes I suppose we should allow to have both for one config because it could be needed by users (repo with dp and side files to validate for example). And now it's nicely synced with `nested` preset of `goodtables` which in a future could allow us support more types of data for one report validation.
- so I've created #169 to discuss concrete mechanisms

File: goodtablesio/integrations/s3/tasks/jobconf.py (hunk 3 above)
Comment (roll) [inline goodtablesio/integrations/s3/tasks/jobconf.py]: @amercader 
With condition I suppose we was missing `raise` at first place so it should be:

```python
except botocore.exceptions.ClientError as exception:
    if exception.response['Error']['Code'] == 'NoSuchKey':
        job_conf_text = None
    raise
```

**Pull request merged**